examples/benchmark/utils.py

```diff
@@ -0,0 +1,206 @@
+
+    Optionally, similarity measures between detected and annotated
+    RR-intervals can be calculated. As this requires interpolation, it may
+    take some time for long signals. Setting a timeout only works on Unix
+    systems.
+
+    Parameters
+    ----------
+    record : ECGRecord
+        As received from `reader_dispatch`.
+    detector : str
+        String identifier of the detector to be used.
+    signal_len : int
+        Length to which the signal should be sliced.
+    max_distance : float
+        Maximum temporal distance in seconds between detected and annotated
+        beats to count as a successful detection.
+    calc_rri_similarity : bool
+        If `True`, calculate similarity measures between detected and
+        annotated RR-intervals (computationally expensive for long
+        signals).
+    timeout : int
+        Number of seconds after which to attempt cancelling execution using
+        SIGALRM.
+    max_timeouts : int
+        Number of timeouts after which a detector is skipped completely.
+
+    Returns
+    -------
+    dict[str, Any]
+        A dictionary containing evaluation results.
+    """
+    signal_len_samples = int(signal_len * record.fs*60)
+    ecg = record.ecg[:signal_len_samples]
+    annotation = record.annotation[record.annotation < signal_len_samples]
+    fs = int(record.fs)
+
+    try:
+        if _timeouts[detector] >= max_timeouts:
+            raise RuntimeError(f'{detector} already has {max_timeouts} timeouts, skipping.')
+
+        if os.name == 'posix':
+            signal.alarm(timeout)
+        start = time.perf_counter()
+        detection = _detector_dispatch(ecg, fs, detector)
+        runtime = time.perf_counter()-start
+        TP, FP, FN = sleepecg.compare_heartbeats(
+            detection,
+            annotation,
+            int(max_distance * record.fs),
+        )
+        if calc_rri_similarity:
+            rri_similarity = sleepecg.rri_similarity(detection, annotation)
+
+        error_message = ''
+
+    except Exception as error:
+        if isinstance(error, TimeoutError):
+            _timeouts[detector] += 1
+            print(f'Timeout: {detector}, {record.id}:{record.lead}')
+        runtime = np.nan
+        TP = []
+        FP = []
+        FN = annotation
+        error_message = repr(error)
+
+    finally:
+        if os.name == 'posix':
+            signal.alarm(0)
+
+    return {
+        'record_id': record.id,
+        'lead': record.lead,
+        'fs': record.fs,
+        'num_samples': len(ecg),
+        'detector': detector,
+        'max_distance': max_distance,
+        'runtime': runtime,
+        'TP': len(TP),
+        'FP': len(FP),
+        'FN': len(FN),
+        'pearsonr': rri_similarity.pearsonr if calc_rri_similarity else np.nan,
+        'spearmanr': rri_similarity.spearmanr if calc_rri_similarity else np.nan,
+        'rmse': rri_similarity.rmse if calc_rri_similarity else np.nan,
```

Comment:
I wonder if we should only include those entries `if calc_rri_similarity` is `True`. Or is it important to always have exactly the same set of dict keys?